### ISSUE TITLE
Recording videos

### ISSUE THREAD

**Issue by wojtulab** [EliasKotlyar/Xiaomi-Dafang-Hacks#106]
Hello,
is it possible to implement catch image/rtsp/live view during recording in background?
It would be very handy. I could view my camera live in any time and watch archived videos via ftp.

**Comment by kszere** [comment]
When v4l2rtspserver-master is running, you can record with avconv.

**Comment by wojtulab** [comment]
locally to sdcard?

**Comment by kszere** [comment]
Yes to microSD or with motion from outside
.

**Comment by wojtulab** [comment]
great but what does "motion" mean?

**Comment by kszere** [comment]
[https://motion-project.github.io/index.html](https://motion-project.github.io/index.html)

**Comment by wojtulab** [comment]
are u sure with avconv?
```
[root@dafang:~]# avconv
-sh: avconv: not found
[root@dafang:~]# busybox avconv
avconv: applet not found
```

**Comment by EliasKotlyar** [comment]
try /system/sdcard/bin/avconf

**Comment by kszere** [comment]
@wojtulab Before use, please [download](https://github.com/EliasKotlyar/Xiaomi-Dafang-Hacks/files/1654102/avconv.zip)  and put file to `/system/sdcard/bin/avconf`

**Comment by wojtulab** [comment]
downloaded, unpacked and...
```
[root@dafang:~]# /system/sdcard/bin/avconv
/system/sdcard/bin/avconv: line 1: can't open ▒▒▒%: no such file
/system/sdcard/bin/avconv: line 1:ELF▒%: not found
/system/sdcard/bin/avconv: line 4: syntax error: unexpected "("
```

**Comment by shinebar1001** [comment]
Looks like you uploaded the binary in ASCII mode - can you please doublecheck that? I just tried the binary, it is working.
`[root@DAFANGHACKS:tmp]# ./avconv
avconv version 13_dev0, Copyright (c) 2000-2018 the Libav developers
  built on Jan 23 2018 00:51:16 with gcc 5.4.0 (Ubuntu 5.4.0-6ubuntu1~16.04.1) 20160609
Hyper fast Audio and Video encoder
usage: avconv [options] [[infile options] -i infile]... {[outfile options] outfile}...

Use -h to get full help or, even better, run 'man avconv'
`

**Comment by wojtulab** [comment]
Yes you are right. Switched to binary mode and another issue:
`avconv version 13_dev0, Copyright (c) 2000-2018 the Libav developers
  built on Jan 23 2018 00:51:16 with gcc 5.4.0 (Ubuntu 5.4.0-6ubuntu1~16.04.1) 20160609
[rtsp @ 0x17f98a0] max delay reached. need to consume packet
[rtsp @ 0x17f98a0] RTP: missed 115 packets
[rtsp @ 0x17f98a0] max delay reached. need to consume packet
[rtsp @ 0x17f98a0] RTP: missed 189 packets
[rtsp @ 0x17f98a0] max delay reached. need to consume packet
[rtsp @ 0x17f98a0] RTP: missed 54 packets
[rtsp @ 0x17f98a0] max delay reached. need to consume packet
[rtsp @ 0x17f98a0] RTP: missed 9 packets
[rtsp @ 0x17f98a0] max delay reached. need to consume packet
[rtsp @ 0x17f98a0] RTP: missed 50 packets
[rtsp @ 0x17f98a0] max delay reached. need to consume packet
[rtsp @ 0x17f98a0] RTP: missed 10 packets
[rtsp @ 0x17f98a0] Non-increasing DTS in stream 0: packet 14 with DTS 83447, packet 15 with DTS 81681
[rtsp @ 0x17f98a0] Non-increasing DTS in stream 1: packet 66 with DTS 38822, packet 67 with DTS 34994`

**Comment by kszere** [comment]
#### With audio
```./avconv -rtsp_transport tcp -y -i rtsp://0.0.0.0:8554/unicast -vcodec copy -c:a aac -strict experimental -t 10 /system/sdcard/DCIM/VideoRecorder/`date +%Y-%m-%d_%H.%M`.mov```

#### Without audio
```./avconv -rtsp_transport tcp -y -i rtsp://0.0.0.0:8554/unicast -vcodec copy -an -t 60 /system/sdcard/DCIM/VideoRecorder/`date +%Y-%m-%d_%H.%M`.mjpeg </dev/null >/dev/null 2>/tmp/cam01.log &```

**Comment by wojtulab** [comment]
ok, thank you - how to stop/save it?
I've tried with ctrl+c but there is no output file...

`/system/sdcard/bin/avconv -rtsp_transport tcp -y -i rtsp://192.168.0.8:8554/unicast -vcodec copy -c:a aac -strict experimental -t 10 /system/sdcard/nagranie.mov
avconv version 13_dev0, Copyright (c) 2000-2018 the Libav developers
  built on Jan 23 2018 00:51:16 with gcc 5.4.0 (Ubuntu 5.4.0-6ubuntu1~16.04.1) 20160609
[rtsp @ 0x191f8e0] CSeq 5 expected, 0 received.
^C  Last message repeated 193 times
`

**Comment by kszere** [comment]
@wojtulab Pojawia Ci się błąd [rtsp @ 0x191f8e0] CSeq 5 expected, 0 received. Jeszcze przed utworzeniem pliku.

You receive return with error  [rtsp @ 0x191f8e0] CSeq 5 expected, 0 received before create file.

Spróbuj tak: / Try:
```./avconv -rtsp_transport tcp -y -i rtsp://0.0.0.0:8554/unicast -vcodec copy -c:a aac -strict experimental -t 60  /system/sdcard/`date +%Y-%m-%d_%H.%M`.mjpeg```

This create 60 sec video.
To utworzy plik wideo z 10 sekundowym nagraniem.

**Comment by wojtulab** [comment]
@kszere hmm... I am afraid that I've got the same result:

```
avconv version 13_dev0, Copyright (c) 2000-2018 the Libav developers
  built on Jan 23 2018 00:51:16 with gcc 5.4.0 (Ubuntu 5.4.0-6ubuntu1~16.04.1) 20160609
[rtsp @ 0x18168e0] CSeq 5 expected, 0 received.
    Last message repeated 71 times
```

dzieki za pomoc tak w ogole! :)

**Comment by kszere** [comment]
![image](https://user-images.githubusercontent.com/7155051/36567557-2296a096-1827-11e8-8635-686b46613fd1.png)

Używasz cfw-1.3.bin? Ja używam cfw-1.2.bin. Nie wiem czy ma to znacznie w tym przypadku. U mnie zapisuje nagranie poprawnie. :)

You use cfw-1.3.bin? I use cfw-1.2.bin. I not know do have meaning in this situation. For me recordings proplely.

**Comment by wojtulab** [comment]
I'm not sure... 1.2 CFW , I guess...

**Comment by don86nl** [comment]
> With audio
> ./avconv -rtsp_transport tcp -y -i rtsp://0.0.0.0:8554/unicast -vcodec copy -c:a aac -strict experimental -t 10 /system/sdcard/DCIM/VideoRecorder/`date +%Y-%m-%d_%H.%M`.mov

Is there a way to save it to a NAS, rather then the SD card from the camera?
The NAS is accessible from FTP or SMB (//0.0.0.0/Public)

**Comment by shinebar1001** [comment]
you can upload the file with curl
`/system/sdcard/bin/curl --upload-file /system/sdcard/DCIM/VideoRecorder/date+%Y-%m-%d_%H.%M.mov  -u username:password ftp://sitename.com/directory/myfile`

**Comment by wojtulab** [comment]
I found that **avconv** radomly can record but generaly receives error. 
```
avconv version 13_dev0, Copyright (c) 2000-2018 the Libav developers
  built on Jan 23 2018 00:51:16 with gcc 5.4.0 (Ubuntu 5.4.0-6ubuntu1~16.04.1) 20160609
[rtsp @ 0x18168e0] CSeq 5 expected, 0 received.
    Last message repeated 71 times
```
How can I resolve it?

**Comment by shinebar1001** [comment]
just put the date into a variable:
``FILE=`date +%Y-%m-%d_%H.%M.mov` &&
./avconv -rtsp_transport tcp -y -i rtsp://0.0.0.0:8554/unicast -vcodec copy -c:a aac -strict experimental -t 10 /system/sdcard/DCIM/VideoRecorder/$FILE &&
/system/sdcard/bin/curl --upload-file /system/sdcard/DCIM/VideoRecorder/$FILE -u username:password ftp://sitename.com/directory/$FILE
``

**Comment by don86nl** [comment]
Ah, nice! thanks! (Pretty new into programming)

**Comment by nobet78** [comment]
@kszere "Yes to microSD or with motion from outside" 
Jak to zrobić żeby nagrywało się dopiero po wykryciu ruchu?
How to do it to record only after motion detection?

**Comment by kszere** [comment]
@nobet78 nie znam sposobu, aby zrobić to na samej kamerce.
Może dałoby się poprzez robienie co sekundę zrzut klatki i porównywanie zmian, lecz nigdy to nie będzie 1:1 takie samo. Trzeba porównywać zmianę pixeli w zakresie n% dopuszczalnym.

I not have solution now.

**Comment by nobet78** [comment]
@kszere możesz mi pomóc? chciałem skorzystać z avconv mając nadzieję że działa w pętli ale program przestaje działać po pierwszych 60 sekundach, czy jest jakiś sposób żeby nagrywanie za pomocą avconv trwało bez końca?

can you help me? I wanted to use avconv, hoping that it works in a loop, but the program stops working after the first 60 seconds, is there any way that avconv recording forever?

**Comment by wojtulab** [comment]
Hello,
I found some issue during recording with avconv.... 

When I add paremeter '-A' to /system/sdcard/bin/v4l2rtspserver-master

avconv can record it normally 
```avconv version 13_dev0, Copyright (c) 2000-2018 the Libav developers
  built on Jan 23 2018 00:51:16 with gcc 5.4.0 (Ubuntu 5.4.0-6ubuntu1~16.04.1) 20160609
[rtsp @ 0x1b908a0] Estimating duration from bitrate, this may be inaccurate
Input #0, rtsp, from 'rtsp://wk:Lipiec28@0.0.0.0:8554/unicast':
  Metadata:
    title           : LIVE555 Streaming Media v2017.10.28
    comment         : LIVE555 Streaming Media v2017.10.28
  Duration: N/A, start: 0.080300, bitrate: N/A
    Stream #0:0: Video: h264 (Constrained Baseline)
      yuv420p, 1280x720
      17.08 fps, 90k tbn
Output #0, mov, to '/system/sdcard/test2.mov':
  Metadata:
    title           : LIVE555 Streaming Media v2017.10.28
    comment         : LIVE555 Streaming Media v2017.10.28
    encoder         : Lavf58.1.0
    Stream #0:0: Video: h264 (Constrained Baseline) [avc1 / 0x31637661]
      yuv420p, 1280x720, q=2-31
      90k tbn
Stream mapping:
  Stream #0:0 -> #0:0 (copy)
Press ctrl-c to stop encoding

```

but when the audio is running (without parameter -A) it crashes everytime.
`avconv version 13_dev0, Copyright (c) 2000-2018 the Libav developers
  built on Jan 23 2018 00:51:16 with gcc 5.4.0 (Ubuntu 5.4.0-6ubuntu1~16.04.1) 20160609
[rtsp @ 0x18168e0] CSeq 5 expected, 0 received.
    Last message repeated 71 times`

**Comment by nik0** [comment]
Hello,
For me avconv never exists. I start v4l2rtspserver-master with the -A option, and I get these traces:
`/system/sdcard/bin/avconv -v debug -rtsp_transport tcp -y -i rtsp://...:...@192.168.0.18:8555/unicast -vcodec copy  -strict experimental -t 10 /system/sdcard/file.mkv
avconv version 13_dev0, Copyright (c) 2000-2018 the Libav developers
  built on Jan 23 2018 00:51:16 with gcc 5.4.0 (Ubuntu 5.4.0-6ubuntu1~16.04.1) 20160609
  configuration: --target-os=linux --arch=mipsel --enable-cross-compile --cross-prefix=mipsel-linux-gnu- --prefix=/opt/cross --extra-ldflags=-static --enable-nonfree --disable-doc --enable-gpl
  libavutil     56.  7. 0 / 56.  7. 0
  libavcodec    58.  8. 0 / 58.  8. 0
  libavformat   58.  1. 0 / 58.  1. 0
  libavdevice   57.  0. 2 / 57.  0. 2
  libavfilter    7.  0. 0 /  7.  0. 0
  libavresample  4.  0. 0 /  4.  0. 0
  libswscale     5.  0. 0 /  5.  0. 0
Splitting the commandline.
Reading option '-v' ... matched as option 'v' (set libav* logging level) with argument 'debug'.
Reading option '-rtsp_transport' ... matched as AVOption 'rtsp_transport' with argument 'tcp'.
Reading option '-y' ... matched as option 'y' (overwrite output files) with argument '1'.
Reading option '-i' ... matched as input file with argument 'rtsp://...:...@192.168.0.18:8555/unicast'.
Reading option '-vcodec' ... matched as option 'vcodec' (force video codec ('copy' to copy stream)) with argument 'copy'.
Reading option '-strict' ... matched as AVOption 'strict' with argument 'experimental'.
Reading option '-t' ... matched as option 't' (record or transcode "duration" seconds of audio/video) with argument '10'.
Reading option '/system/sdcard/file.mkv' ... matched as output file.
Finished splitting the commandline.
Parsing a group of options: global .
Applying option v (set libav* logging level) with argument debug.
Applying option y (overwrite output files) with argument 1.
Successfully parsed a group of options.
Parsing a group of options: input file rtsp://...:...@192.168.0.18:8555/unicast.
Successfully parsed a group of options.
Opening an input file: rtsp://...:...@192.168.0.18:8555/unicast.
[rtsp @ 0x19d18e0] SDP:
v=0
o=- 1521028109927473 1 IN IP4 192.168.0.18
s=LIVE555 Streaming Media v2017.10.28
i=LIVE555 Streaming Media v2017.10.28
t=0 0
a=tool:LIVE555 Streaming Media v2017.10.28
a=type:broadcast
a=control:*
a=range:npt=0-
a=x-qt-text-nam:LIVE555 Streaming Media v2017.10.28
a=x-qt-text-inf:LIVE555 Streaming Media v2017.10.28
m=video 0 RTP/AVP 96
c=IN IP4 0.0.0.0
b=AS:168
a=rtpmap:96 H264/90000
a=fmtp:96 profile-level-id=ffffc01f;sprop-parameter-sets=Z0LAH5WoFAFuhAAAAwAEAAADAMgQ,aO48gA==
a=x-dimensions:1280,720
a=control:track1

[rtsp @ 0x19d18e0] video codec set to: h264
[rtsp @ 0x19d18e0] Extradata set to 0x19d48c0 (size: 33)
[rtsp @ 0x19d18e0] setting jitter buffer size to 0
[rtsp @ 0x19d18e0] hello state=0
[h264 @ 0x19d44e0] nal_unit_type: 7, nal_ref_idc: 3
[h264 @ 0x19d44e0] nal_unit_type: 8, nal_ref_idc: 3
[h264 @ 0x19d44e0] nal_unit_type: 7, nal_ref_idc: 3
[h264 @ 0x19d44e0] nal_unit_type: 8, nal_ref_idc: 3
[h264 @ 0x19d44e0] nal_unit_type: 5, nal_ref_idc: 3
[h264 @ 0x19d44e0] Format yuv420p chosen by get_format().
[h264 @ 0x19d44e0] Reinit context to 1280x720, pix_fmt: 0
[h264 @ 0x19d44e0] no picture
[h264 @ 0x19d44e0] nal_unit_type: 1, nal_ref_idc: 2
[h264 @ 0x19d44e0] no picture
[h264 @ 0x19d44e0] nal_unit_type: 1, nal_ref_idc: 2
[h264 @ 0x19d44e0] no picture
[h264 @ 0x19d44e0] nal_unit_type: 1, nal_ref_idc: 2
[h264 @ 0x19d44e0] no picture
[h264 @ 0x19d44e0] nal_unit_type: 1, nal_ref_idc: 2
[h264 @ 0x19d44e0] no picture
[h264 @ 0x19d44e0] nal_unit_type: 1, nal_ref_idc: 2
    Last message repeated 5 times
[rtsp @ 0x19d18e0] CSeq 6 expected, 0 received._
`

any idea ?
thanks,
Nik0

**Comment by herrxyz** [comment]
is there a reason why avconv is not in /bin by default?

**Comment by jmtatsch** [comment]
@herrxyz i guess its file size is too large at 51,6 mb

**Comment by Darkblade48** [comment]
I am also encountering a similar problem.

I am trying to run the following:

`avconv -rtsp_transport tcp -y -i rtsp://username:password@192.168.2.115:8554/unicast -vcodec copy -an -t 60 /system/sdcard/DCIM/`date +%Y-%m-%d_%H.%M`.mjpeg`

However, I just get a 'Killed' message. For whatever reason, it also stops my rtsp-h264 service (it was previously running, with a proper PID, but after running this command via SSH, it makes it NOK).

**Comment by jmtatsch** [comment]
Your system runs out of memory so it kills something to regain memory. Try lowering the resolution.

**Comment by Darkblade48** [comment]
@jmtatsch I am using the flag -s qvga (320x240) and now get the following before the process is killed


> Non-increasing DTS in stream 1: packet 14 with DTS 49480, packet 15 with DTS 35374
[rtsp @ 0x18380e0] Non-increasing DTS in stream 1: packet 22 with DTS 67851, packet 23 with DTS 53745
[rtsp @ 0x18380e0] Non-increasing DTS in stream 1: packet 37 with DTS 103923, packet 38 with DTS 89817
[rtsp @ 0x18380e0] Non-increasing DTS in stream 1: packet 45 with DTS 121903, packet 46 with DTS 107797
[rtsp @ 0x18380e0] Non-increasing DTS in stream 1: packet 60 with DTS 157630, packet 61 with DTS 143524

**Comment by Shaker1978** [comment]
Is there any solution so far?

I get the same errors of avconv about Non-increasing DTS or the CSeq 5 expected. No output file is being created.

**Comment by Darkblade48** [comment]
> Is there any solution so far?
> 
> I get the same errors of avconv about Non-increasing DTS or the CSeq 5 expected. No output file is being created.

I outlined what I did to get it working in #452

**Comment by Shaker1978** [comment]
Alright, from what I understood, we have to set the resolution down to 640x360? That's not what I want. What did I buy a 1080p cam for then?
I mean, 1080p isn't working anyway, but at lease I would want to record the currently set 1600x900 resolution...

**Comment by Darkblade48** [comment]
> 
> 
> Alright, from what I understood, we have to set the resolution down to 640x360? That's not what I want. What did I buy a 1080p cam for then?
> I mean, 1080p isn't working anyway, but at lease I would want to record the currently set 1600x900 resolution...

Yes, that's pretty much it. If you want to record [b]and[/b] stream at the same time, the camera doesn't have enough processing power to do it at full 1080p. 

If you only want to do one or the other, then yes, it is possible.

**Comment by Hussamfhaily** [comment]
why does homebridge crash when I press record video button?

**Comment by thisisjerk** [comment]
Would it overwrite the most earliest videos once the SDCard is full?

**Comment by Ch1zh** [comment]
Ping. I would join to the question of overwriting the files on SD card..
